File: core-algorithms/3_sort/2_quick_sort/quick_sort.ts
const quickSort = (arr: number[]): number[] => {
  if (arr.length <= 1) {
    return arr;
  }

  // Picking the middle element helps avoid O(n^2) on sorted arrays
  const pivotIndex = Math.floor(arr.length / 2);
  const pivot = arr[pivotIndex];

  const lhs = arr.filter((val) => val < pivot);
  const middle = arr.filter((val) => val === pivot);
  const rhs = arr.filter((val) => val > pivot);

  return [...quickSort(lhs), ...middle, ...quickSort(rhs)];
};

const quickSortedArray = quickSort([2, 43, 1, 23, 54, 13, 119, 0]);

console.log(quickSortedArray);

File: core-algorithms/3_sort/2_quick_sort/quick_sort.py
import math
def quickSort(lis):

    if(len(lis) <= 1):
        return lis

    medianElement = lis[math.floor(len(lis)/2)]
    leftList = list(filter(lambda x : x < medianElement, lis))
    middleList = list(filter(lambda x : x == medianElement, lis))
    rightList = list(filter(lambda x : x > medianElement, lis))
    return [*quickSort(leftList), *middleList, *quickSort(rightList)]
    

quickSortedArray = quickSort([2, 43, 1, 23, 54, 13, 119, 0]);

print(quickSortedArray);